JavaParser refactoring

sebkur opened the pull request:
This PR is the result of the discussion on this one: https://github.com/javaparser/javaparser/pull/207

The main parsing methods of the JavaParser now do not return a Node
subclass anymore, but instead an instance of the new ParseResult class.
This class wraps a Node plus some extra information about the parsing
process. At the moment the additional information is the list of tokens
that the parser has built while parsing the source code. It could
however be extended in the future to provide more information about the
parsing process.

Changes in detail:

Enabled the COMMON_TOKEN_ACTION=true option in the java_1_8.jj file.

Implemented a CommonTokenAction(Token token) method in the
TOKEN_MGR_DECLS that collects encountered tokens in a list.

Added a getter for this list of tokens in the ASTParser.

Added the ParseResult class as a wrapper for Node and List<Token>.

Changed parsing methods of JavaParser to return ParseResults instead of
Node subclasses.

Fixed testing source code to work with the new JavaParser API.

Cleaned up the JavaParser's parse() methods that parse a
CompilationUnit:
- Consistent overloaded versions for all input types (String,
  InputStream, File and Reader)
- Only the parse() method that accepts a String as input and mandatory
  'encoding' and 'considerComments' arguments contains the logic of how
  to actually parse and create a result. All other version just forward
  to this method by building a String and passing it to this method.
- Added JavaDoc for some undocumented public methods.
- Fixed JavaDoc: new return types.

sebkur committed:
JavaParser refactoring

The main parsing methods of the JavaParser now do not return a Node
subclass anymore, but instead an instance of the new ParseResult class.
This class wraps a Node plus some extra information about the parsing
process. At the moment the additional information is the list of tokens
that the parser has built while parsing the source code. It could
however be extended in the future to provide more information about the
parsing process.

Changes in detail:

Enabled the COMMON_TOKEN_ACTION=true option in the java_1_8.jj file.

Implemented a CommonTokenAction(Token token) method in the
TOKEN_MGR_DECLS that collects encountered tokens in a list.

Added a getter for this list of tokens in the ASTParser.

Added the ParseResult class as a wrapper for Node and List<Token>.

Changed parsing methods of JavaParser to return ParseResults instead of
Node subclasses.

Fixed testing source code to work with the new JavaParser API.

Cleaned up the JavaParser's parse() methods that parse a
CompilationUnit:
* Consistent overloaded versions for all input types (String,
  InputStream, File and Reader)
* Only the parse() method that accepts a String as input and mandatory
  'encoding' and 'considerComments' arguments contains the logic of how
  to actually parse and create a result. All other version just forward
  to this method by building a String and passing it to this method.
* Added JavaDoc for some undocumented public methods.
* Fixed JavaDoc: new return types.

ftomassetti commented on javaparser-core/src/main/java/com/github/javaparser/JavaParser.java:
```diff
@@ -86,72 +83,145 @@ public static void setDoNotAssignCommentsPreceedingEmptyLines(boolean doNotAssig
      */
-    public static CompilationUnit parse(final InputStream in,
-                                        final String encoding, boolean considerComments) throws ParseException {
+    public static ParseResult<CompilationUnit> parse(final String code,
```
To maintain compatibility we could have both one method with the same name and the same return type as before and a new method named "parseWithTokens" or something similar. The new method would return the ParseResult and the method with the old name ("parse") would just call the new method, take the node and throw away the tokens. What do you think? I know it could be a bit repetitive to write but in this way we would stay compatible and offer a clean interface for many users which are not interested in tokens, while giving all the information to users interested in both nodes and tokens

sebkur committed:
JavaParser: Renamed methods with new return type

Renamed the methods that returned a ParseResult instead of a Node subclass
after the previous commit. The are now called 'parseToResult' and
'parse*ToResult'.

Recreated the former method signatures by forwarding to the
parseToResult methods and unpacking the Node from the ParseResult as
return value.

Fixed JavaDoc return type information.

sebkur committed:
Merge branch 'master' into new_parser_api
